src/fontations/lib.rs

```diff
@@ -768,6 +780,83 @@ extern "C" fn _hb_fontations_paint_glyph(
     let _ = color_glyph.paint(location, &mut painter);
 }
 
+extern "C" fn _hb_fontations_glyph_name(
+    _font: *mut hb_font_t,
+    font_data: *mut ::std::os::raw::c_void,
+    glyph: hb_codepoint_t,
+    name: *mut ::std::os::raw::c_char,
+    size: ::std::os::raw::c_uint,
+    _user_data: *mut ::std::os::raw::c_void,
+) -> hb_bool_t {
+    let data = unsafe { &mut *(font_data as *mut FontationsData) };
+
+    let glyph_name = data.glyph_names.get(GlyphId::new(glyph));
+    match glyph_name {
+        None => false as hb_bool_t,
+        Some(glyph_name) => {
+            let glyph_name = glyph_name.as_str();
+            // Copy the glyph name into the buffer, up to size-1 bytes
+            let len = glyph_name.len().min(size as usize - 1);
+            unsafe {
+                std::ptr::copy_nonoverlapping(glyph_name.as_ptr(), name as *mut u8, len);
+                *name.add(len) = 0;
+            }
+            true as hb_bool_t
+        }
+    }
+}
+
+extern "C" fn _hb_fontations_glyph_from_name(
+    _font: *mut hb_font_t,
+    font_data: *mut ::std::os::raw::c_void,
+    name: *const ::std::os::raw::c_char,
+    len: ::std::os::raw::c_int,
+    glyph: *mut hb_codepoint_t,
+    _user_data: *mut ::std::os::raw::c_void,
+) -> hb_bool_t {
+    let data = unsafe { &mut *(font_data as *mut FontationsData) };
+
+    let name = unsafe { std::slice::from_raw_parts(name as *const u8, len as usize) };
+    let name = std::str::from_utf8(name).unwrap_or_default();
+
+    let glyph_from_names = data.glyph_from_names.load(Ordering::Acquire);
+    let glyph_from_names = if glyph_from_names.is_null() {
+        loop {
+            let mut map = HashMap::new();
+            for (glyph_id, glyph_name) in data.glyph_names.iter() {
+                map.insert(glyph_name.to_string(), u32::from(glyph_id));
+            }
+            let map = Box::into_raw(Box::new(map));
+            if data.glyph_from_names.compare_exchange(
+                null_mut(),
+                map,
+                Ordering::Acquire,
+                Ordering::Relaxed,
+            ) == Ok(null_mut())
+            {
+                break map;
+            } else {
+                unsafe {
+                    let _ = Box::from_raw(map);
+                }
+            }
+        }
+    } else {
+        glyph_from_names
+    };
+
+    let glyph_id = unsafe { &*glyph_from_names }.get(name);
```

Comment:
with the above change, this entire block becomes
```rust
    let glyph_from_names = data.glyph_from_names.get_or_init(|| {
        data.glyph_names
            .iter()
            .map(|(gid, name)| (name.to_string(), gid.to_u32()))
            .collect()
    });
    let glyph_id = glyph_from_names.get(name);
```

Reply:
Oof. Thanks.

Reply:
I tried to do the same to the other use of Atomic I had but I gave up. I think this is good enough for now.